ci-operator/step-registry/ipi/install: Use spot compute on AWS

wking opened the pull request:
Via [`spotMarketOptions`][1], supported [since][3] [4.5][2].  We'll fail CI sometimes on this, if there are no spots available.  But we fail CI sometimes on no on-demand instances being available, and we can always retest.  The cost savings of spot instances are expected to outweigh the reduced CI success rates.

[1]: https://docs.openshift.com/container-platform/4.10/machine_management/creating_machinesets/creating-machineset-aws.html#machineset-creating-non-guaranteed-instance_creating-machineset-aws
[2]: https://docs.openshift.com/container-platform/4.5/machine_management/creating_machinesets/creating-machineset-aws.html#machineset-creating-non-guaranteed-instance_creating-machineset-aws
[3]: https://docs.openshift.com/container-platform/4.5/release_notes/ocp-4-5-release-notes.html#ocp-4-5-aws-machinesets-support-spot-instances

wking commented:
Most of the rehearsals are irrelevant, because the rehearsal engine is understandably not going to look at my shell and notice the AWS switching.  But [the e2e-operator run][1] confirmed this working:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/30937/rehearse-30937-pull-ci-openshift-aws-load-balancer-operator-main-e2e-aws-operator/1553097666978648064/artifacts/e2e-aws-operator/ipi-install-install/build-log.txt | grep Patched
Patched spotMarketOptions into /tmp/installer/openshift/99_openshift-cluster-api_worker-machineset-0.yaml
Patched spotMarketOptions into /tmp/installer/openshift/99_openshift-cluster-api_worker-machineset-1.yaml

$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/30937/rehearse-30937-pull-ci-openshift-aws-load-balancer-operator-main-e2e-aws-operator/1553097666978648064/artifacts/e2e-aws-operator/gather-extra/artifacts/machinesets.json | jq -r '.items[] | (.status.availableReplicas | tostring) + " " + (.spec.template.spec.providerSpec.value | {spotMarketOptions} | tostring)'
2 {"spotMarketOptions":{}}
1 {"spotMarketOptions":{}}
```

At least right now [in us-east-2][2].  Unclear how reliably we'd be able to get spots in our target zone/instance-type requests, so I've moved the pull request to draft to keep folks from landing it without the discussion and consensus-building we'd want for a change like this.

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/30937/rehearse-30937-pull-ci-openshift-aws-load-balancer-operator-main-e2e-aws-operator/1553097666978648064
[2]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/30937/rehearse-30937-pull-ci-openshift-aws-load-balancer-operator-main-e2e-aws-operator/1553097666978648064#1:build-log.txt%3A30

JoelSpeed reviewed (commented):
Code wise this looks fine to me. As it's only workers I'm not hugely concerned about the tests either. The only issue I can foresee is that, in the Machine API e2e, we have tests that explicitly enable spot instances, that's the e2e-aws-operator job, if we were to enable this there, then it may break those jobs (we aren't currently).

But yeah, I'm +1 to this in general for MAPA and if others are happy would be ok with it being rolled out further too

wking commented:
[This rehearsal][1] would be the one with spot, and it failed a number of test-cases.  But spot itself seemed to work out:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/30937/rehearse-30937-pull-ci-openshift-machine-api-provider-aws-main-e2e-aws/1554236456036208640/artifacts/e2e-aws/gather-extra/artifacts/machinesets.json | jq -r '.items[] | (.status.availableReplicas | tostring) + " " + (.spec.template.spec.providerSpec.value | {spotMarketOptions} | tostring)'
2 {"spotMarketOptions":{}}
1 {"spotMarketOptions":{}}
```

With backing Machines that... did not last the whole run?

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/30937/rehearse-30937-pull-ci-openshift-machine-api-provider-aws-main-e2e-aws/1554236456036208640/artifacts/e2e-aws/gather-extra/artifacts/machines.json | jq -r '.items[].metadata | .creationTimestamp + " " + .name' | sort
2022-08-01T23:07:42Z ci-op-3h8wi78t-db23b-dtqqv-master-0
2022-08-01T23:07:42Z ci-op-3h8wi78t-db23b-dtqqv-master-1
2022-08-01T23:07:42Z ci-op-3h8wi78t-db23b-dtqqv-master-2
2022-08-01T23:11:25Z ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1b-kg8r6
2022-08-02T00:32:30Z ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-49jfp
2022-08-02T00:32:35Z ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-l76qr
```

And yeah, spot issues:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/30937/rehearse-30937-pull-ci-openshift-machine-api-provider-aws-main-e2e-aws/1554236456036208640/artifacts/e2e-aws/gather-extra/artifacts/events.json | jq -r '[.items[] | select(tostring | contains("worker-us-west-1a")) | .ts = .firstTimestamp // .metadata.creationTimestamp] | sort_by(.ts)[] | .ts + " " + .involvedObject.name + " " + .reason + ": " + .message'
2022-08-01T23:12:08Z ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-48v4l Create: Created Machine ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-48v4l
2022-08-01T23:12:10Z ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-gmhl5 Create: Created Machine ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-gmhl5
2022-08-01T23:12:31Z ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-48v4l Update: Updated Machine ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-48v4l
2022-08-01T23:12:42Z ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-gmhl5 Update: Updated Machine ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-gmhl5
2022-08-01T23:32:45Z ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-48v4l MachineDeleted: Machine openshift-machine-api/machine-api-termination-handler/ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-48v4l/ip-10-0-141-253.us-west-1.compute.internal has been remediated by requesting to delete Machine object
2022-08-01T23:32:45Z ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-48v4l DrainProceeds: Node drain proceeds
2022-08-01T23:32:46Z ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-gmhl5 MachineDeleted: Machine openshift-machine-api/machine-api-termination-handler/ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-gmhl5/ip-10-0-137-163.us-west-1.compute.internal has been remediated by requesting to delete Machine object
2022-08-01T23:32:48Z ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-nj9db FailedCreate: ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-nj9db: reconciler failed to Create machine: failed to launch instance: error creating EC2 instance: InsufficientInstanceCapacity: There is no Spot capacity available that matches your request.
        status code: 500, request id: 527c33e9-4ece-4c25-8cdc-16b25552f834
2022-08-01T23:32:51Z ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-nj9db FailedCreate: ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-nj9db: reconciler failed to Create machine: failed to launch instance: error creating EC2 instance: InsufficientInstanceCapacity: There is no Spot capacity available that matches your request.
        status code: 500, request id: 178076b9-e8b1-422a-9e20-20b9b95ab3ab
2022-08-01T23:32:55Z ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-nj9db FailedCreate: ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-nj9db: reconciler failed to Create machine: failed to launch instance: error creating EC2 instance: InsufficientInstanceCapacity: There is no Spot capacity available that matches your request.
        status code: 500, request id: a9b55153-2e53-4718-9005-62ccc90c75f2
2022-08-01T23:32:58Z ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-wv9f9 FailedCreate: ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-wv9f9: reconciler failed to Create machine: failed to launch instance: error creating EC2 instance: InsufficientInstanceCapacity: There is no Spot capacity available that matches your request.
        status code: 500, request id: 08b07d21-dcef-4820-99a4-688b0332d9fa
2022-08-01T23:33:01Z ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-nj9db FailedCreate: ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-nj9db: reconciler failed to Create machine: failed to launch instance: error creating EC2 instance: InsufficientInstanceCapacity: There is no Spot capacity available that matches your request.
        status code: 500, request id: 530d95ca-ca98-4154-8d4b-e8f7d0b2986a
2022-08-01T23:33:05Z ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-48v4l DrainRequeued: Node drain requeued: requeue in: 20s
2022-08-01T23:33:05Z ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-gmhl5 DrainProceeds: Node drain proceeds
2022-08-01T23:33:05Z ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-wv9f9 FailedCreate: ci-op-3h8wi78t-db23b-dtqqv-worker-us-west-1a-wv9f9: reconciler failed to Create machine: failed to launch instance: error creating EC2 instance: InsufficientInstanceCapacity: There is no Spot capacity available that matches your request.
        status code: 500, request id: 698ea80c-c445-44e9-a147-35658d08fc37
...
```

Still, a failing presubmit can get retried.  But also, nothing in [the Spyglass page][1] is jumping out at me as "at least some of these test-case failures are likely due to spot availability issues", and you might want something like that in place before moving ahead with this pull request?

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/30937/rehearse-30937-pull-ci-openshift-machine-api-provider-aws-main-e2e-aws/1554236456036208640

JoelSpeed commented:
Lets get this merged and see how it goes, this should unblock setting up spot Cluster Bot clusters as well which will help QE out

/approve
/lgtm

dgoodwin commented:
Looks to have broken some jobs on undefined variable, https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-openshift-release-master-nightly-4.12-e2e-aws-sdn-upgrade/1562023472849227776

Going to revert as standard policy, even though this is surely an easy script fix.

wking commented:
#31577 is taking another run at this.
